content/the-basics-of-go.md

```diff
@@ -0,0 +1,864 @@
+means literally a type of anything), and might be useful when you're dealing with things that you don't know.
+
+But is it couraged to do so? No. You should be careful with interfaces. Go is not a dynamically typed language.
+So you have to validate the given type of the interface.
+
+```go
+func main() {
+	var i interface{}
+
+	i = 42
+	describe(i)
+
+	i = "hello"
+	describe(i)
+}
+
+func describe(i interface{}) {
+	fmt.Printf("(%v, %T)\n", i, i)
+  // Output:
+  // (42, int)
+  // (hello, string)
+}
+```
+
+To validate a given type, there are two ways. One, via the type assertion:
+
+```go
+var i interface{} = "hello"
+
+s := i.(string)
+fmt.Println(s)
+
+s, ok := i.(string)
+fmt.Println(s, ok)
+
+f, ok := i.(float64)
+fmt.Println(f, ok)
+```
+
+Or two, via the type switch:
+
+```go
+func do(i interface{}) {
+	switch v := i.(type) {
+	case int:
+		fmt.Printf("Twice %v is %v\n", v, v*2)
+	case string:
+		fmt.Printf("%q is %v bytes long\n", v, len(v))
+	default:
+		fmt.Printf("I don't know about type %T!\n", v)
+	}
+}
+
+func main() {
+	do(21)
+	do("hello")
+	do(true)
+}
+```
+
+## Error
+
+I want to tell a little story before we go forward. Go's error handling is a bit unique. It's not like C/C++,
+where you have to define a custom error type. But it's not like Java, where you have to define a custom exception.
+There are no try-catches. And there are no exceptions. But Go has an error type that must be defined explicitly,
+which is called `error`, and means your code will be so verbose. I once had an internal talk on one of the
+company that I worked on, and when I explain about why there are `error` on the function signature and how
+verbose your code should be, some PHP and C# developers were confused and immediately don't like the language.
+Not long after, I told them, "well yeah, you could just do panic()". And they laugh, how can a language be so much
+like a developers who would just panic whenever they see an error?
+
+You might don't understand that story, but the point is that you should not be scared or hate the language just
+because its error handling is verbose. After a while, I think being able to handle an error programatically
+(and with that so much verbose-power) is good, because in Go, error are just values. You can decorate the error,
+and find out what was going on and where the error is without trying to read the stack trace.
+
+Here, we're going to use a function from the `strconv` standard library to convert a string into integer.
+
+```go
+package main
+
+import (
+  "strconv"
+  "fmt"
+)
+
+func main() {
+  i, err := strconv.Atoi("420")
+  if err != nil {
+    fmt.Println(err)
+    return
+  }
+
+  fmt.Println(i)
+}
+```
+
+Seems like no problem right, there won't be any error. But what if the input string is not an number, but it's
+a string of "four two oh". It would fail and won't output a valid integer right? That's the error that you as a
+developer should handle on your own.
+
+## Concurrency
```

Comment:
I recommending you to giving bit explanation or referencing about [CSP](http://www.minaandrawos.com/2015/12/06/concurrency-in-golang/), which is the base model of Go's concurrency pattern in order to help they research more for understanding how Go's concurrency works.

Reply:
no